keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -9,105 +9,201 @@
 
-### New API
+### Plugins
+
+- A signer plugin will be a client to the new grpc API thus providing support for external signing.
+- A public key cache will integrate with the same grpc to fetch supported public keys.
+
+### Updates for token generation
+
+As of Kubernetes `v1.13.2`, The Kubernetes Control Plane config accepts `ServiceAccountIssuer` which is an interface of type `serviceaccount.TokenGenerator` as defined in `pkg/serviceaccount/jwt.go`. This enables defining alternate implementations of token generators. Thus, the default TokenGenerator can be replaced by one that integrates with ExternalJWTSigner.
+
+### Updates for supported public keys
+
+Starting ` v1.31` (via. the [PR](https://github.com/kubernetes/kubernetes/pull/125177) by liggit@), `ServiceAccountPublicKeys` in kube-apiserver config was replaced by `ServiceAccountPublicKeysGetter` interface. The static keys held by `ServiceAccountPublicKeys` were hence moved to be managed by the implementation called `StaticPublicKeysGetter`. This presents an opportunity to add a new implementation, one that will get public keys from ExternalJWTSigner depending on how the cluster is configured.
+
+In the same PR, `OpenIDMetadataServer` was updated to accept implementation of `OpenIDMetadataProvider` as defined in  `pkg/serviceaccount/openidmetadata.go`. `OpenIDMetadataProvider` interacts with `ServiceAccountPublicKeysGetter` for serving supported keys which enables flexible replacing `StaticPublicKeysGetter` with a dynamic key getter (will integrate with ExternalJWTSigner) depending on how cluster is configured.
 
-I'm proposing creating a new versioned grpc API under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`
+### Notes/Constraints/Caveats
+
+- A new flag `--key-service-url` will be added to kube-apiserver specifying a unix socket where the key service will be accessible.
+- The flag `--key-service-url` can either be set to the location of a UDS on a filesystem, or be prefixed with an @ symbol and name a UDS in the abstract socket namespace.
+- combining local and external keys will not be supported.
+- Specifying both, `--service-account-key-file` and `--key-service-url` shall result in an error.
+
+### Risks and Mitigations
+
+- New token generation and validation could suffer a performance difference when reading over a socket, as an external process will be signing data.
+- Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket.
+- External signers will only return signature and header while the token will be assembled by TokenGenerator implementation for external signing. This will help kube-apiserver control the payload and thus the claims.
```

Comment:
reword this as a risk that this sentence is the mitigation to (risk is token headers and claims will diverge, mitigation is that kube-apiserver verifies the header algorithm is acceptable, and controls the claims)

Reply:
Done